File: scala/31.scala
import java.util.Arrays
object Solution {
    def nextPermutation(nums: Array[Int]): Unit = {
      if (nums.length == 1) return
      def swap(i: Int, j: Int) {
        val tmp = nums(i)
        nums(i) = nums(j)
        nums(j) = tmp
      }
      var x = 0
      var y = 0
      for (i <- 0 until nums.length)
        for (j <- i until nums.length)
          if (nums(i) < nums(j)) {
            x = i
            y = j
          }
      swap(x, y)
      if (x != y)
        Arrays.sort(nums, x+1, nums.length)
      else
        Arrays.sort(nums)
    }
}

File: go/31.go
import "sort"

func nextPermutation(nums []int) {
	lastFound := -1
	greaterIndex := -1
	for i := range nums {
		index := findMinGreater(nums[i+1:], nums[i])
		if index != -1 {
			lastFound = i
			greaterIndex = index + i + 1
		}
	}
	if lastFound == -1 {
		sort.Ints(nums)
		return
	}

	tmp := nums[lastFound]
	nums[lastFound] = nums[greaterIndex]
	nums[greaterIndex] = tmp
	sort.Ints(nums[lastFound+1:])
}

func findMinGreater(nums []int, target int) int {
	index := -1
	for i := range nums {
		if nums[i] > target {
			if index == -1 || nums[index] > nums[i] {
				index = i
			}
		}
	}
	return index
}